hack/update-godep-licenses.sh

```diff
@@ -70,42 +70,42 @@ process_content () {
   # Start search at package root
   case ${package} in
     github.com/*|golang.org/*|bitbucket.org/*)
-     package_root=$(echo ${package} |awk -F/ '{ print $1"/"$2"/"$3 }')
+     package_root=$(echo "${package}" |awk -F/ '{ print $1"/"$2"/"$3 }')
      ;;
     go4.org/*)
-     package_root=$(echo ${package} |awk -F/ '{ print $1 }')
+     package_root=$(echo "${package}" |awk -F/ '{ print $1 }')
      ;;
     gopkg.in/*)
      # Root of gopkg.in package always ends with '.v(number)' and my contain
      # more than two path elements. For example:
      # - gopkg.in/yaml.v2
      # - gopkg.in/inf.v0
      # - gopkg.in/square/go-jose.v2
-     package_root=$(echo ${package} |grep -oh '.*\.v[0-9]')
+     package_root=$(echo "${package}" |grep -oh '.*\.v[0-9]')
      ;;
     *)
-     package_root=$(echo ${package} |awk -F/ '{ print $1"/"$2 }')
+     package_root=$(echo "${package}" |awk -F/ '{ print $1"/"$2 }')
      ;;
   esac
 
   # Find files - only root and package level
-  local_files=($(
+  mapfile -t local_files < <(
     for dir_root in ${package} ${package_root}; do
       [[ -d ${DEPS_DIR}/${dir_root} ]] || continue
 
       # One (set) of these is fine
-      find ${DEPS_DIR}/${dir_root} \
+      find "${DEPS_DIR}/${dir_root}" \
           -xdev -follow -maxdepth ${find_maxdepth} \
           -type f "${find_names[@]}"
-    done | sort -u))
+    done | sort -u)
 
   local index
   local f
   index="${package}-${type}"
   if [[ -z "${CONTENT[${index}]-}" ]]; then
-    for f in ${local_files[@]-}; do
```

Comment:
Do we have to care if the array is empty? That would make handling this a lot messier

Reply:
I think not, because on the next line there is a `grep` statement looking for copyright or license strings in the files. If the file name is empty, `grep` fails. TBH, I don't know what has been the idea with the original default value. When the array is put into double quotes, the default value means that the `for` loop is run at least once with an empty value (which doesn't work).

Consider this code:

```
#!/bin/bash

a=()

# original behavior
for x in ${a[@]-}; do
    echo "x: $x"
done

# quoting with default value
for y in "${a[@]-}"; do
    echo "y: $y"
done

# quoting without default value
for z in "${a[@]}"; do
    echo "z: $z"
done
```

When run, this produces:

    y:

So, the behavior is supposed to be the same as with the original version, except the script should handle difficult file names (spaces, globbing characters, etc) better after the fix.

Reply:
Good analysis -- I'm not sure that a lot of thought went into the behavior in the first place (or if it did, there are so many ways that Bash arrays can trip you up that I would not be surprised if the current behavior is just an artifact of a misunderstanding of their behavior).

Reply:
So, if `set -o nounset` is on, some bash versions don't allow the for loop over an empty array. This is what bit me in https://github.com/kubernetes/kubernetes/issues/59856. So the default value needs to be there if the array is empty, and it needs to be ignored if it's there:

    # quoting with default value
    for y in "${a[@]-}"; do
        if [[ -z "$y" ]]; then
            continue
        fi
        # do work here
    done